proposed/0006-unified-structured-data-imporvement.md

```diff
@@ -0,0 +1,177 @@
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 3:**
+Logic: Since `previous_owner_keys` in incoming data contains nothing, use `key` = `(sha512(owner_keys), Name)` for look-up.
+Vault will check if new data is sent by valid owner. Use `owner_keys` from the stored data in step 1 to verify signature of data in step 2. If valid replace previous data and store new one:
+```
+Key = (sha512(A-PublicKey), Name)
+Value = StructuredData {
+    other_fields: Value-11,
+    owner_keys: A-PublicKey,
+    previous_owner_keys: NONE,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 4:**
+Client-A wants to transfer ownership of Structured Data to Client-B:
+```
+StructuredData {
+    other_fields: Value-22,
+    owner_keys: B-PublicKey,
+    previous_owner_keys: A-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 5:**
+Logic: Since `previous_owner_keys` in incoming data contains something, use `key` = `(sha512(previous_owner_keys), Name)` for look-up. Calculate a new `key` = `(sha512(owner_keys), Name)` if the update is legal as checked below (as usual).
+Vault will check if new data is sent by valid owner. Use `owner_keys` from the stored data in step 3 to verify signature of data in step 4. If valid replace previous data and store new one:
+```
+Key = (sha512(B-PublicKey), Name)
+Value = StructuredData {
+    other_fields: Value-22,
+    owner_keys: B-PublicKey,
+    previous_owner_keys: A-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 6:**
+Client-A tries to update illegally:
+```
+StructuredData {
+    other_fields: Value-22,
+    owner_keys: Anything,
+    previous_owner_keys: Anything/NONE,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 7:**
+If look-up succeeds, Vault will use owner field from **step 5** == `B-PublicKey` and signature verification fails. So it will not update anything.
+ 
+**step 8:**
+Client-B tries to update:
+```
+StructuredData {
+    other_fields: Value-33,
+    owner_keys: B-PublicKey,
+    previous_owner_keys: NONE,
+    signatures: Signed by B-PrivateKey,
+}
+```
+ 
+**step 9:**
+Logic: Since `previous_owner_keys` in incoming data contains nothing, use `key` = `(sha512(owner_keys), Name)` for look-up.
+Vault will check if new data is sent by valid owner. Use `owner_keys` from the stored data in step 5 to verify signature of data in step 8. If valid replace previous data and store new one:
+```
+Key = (sha512(B-PublicKey), Name)
+Value = StructuredData {
+    other_fields: Value-33,
+    owner_keys: B-PublicKey,
+    previous_owner_keys: NONE,
+    signatures: Signed by B-PrivateKey,
+}
+```
+## Change in the Definition of GETs
+The above shows the cases for PUTs, POSTs and DELETEs. There will be a change in the signature of GETs to achieve the above no-collision scheme. GETs for StructuredData need to be changed from:
+```
+GET(Name, DataRequest::StructuredData(u64))
+```
+to
+```
+GET(Name, DataRequest::StructuredData(SHA512(PublicKeys)))
+```
```

Comment:
This was with Reference to the now outdated `DataRequest` as it now also contains `NameType` field. I'll update this section. But I still have some questions if you could answer. So present day `DataRequest`s are more like `DataRequest::StructuredData(id, tag)` .. when we route the `GET`s we already are calculating the `name` as `sha512(id + tag)` , so why would you additionally require tags `GET`s ? In `PUT/POST/DELETE`s you do have the entire SD with you as a payload and can thus extract the tag information if you need to and use that for some logic. So i don't get what you would miss `u64` was not retained.